### PR TITLE
util: record sqls and heap profile when memory usage is higher than 80% system memory. (#18858)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #18858 to release-4.0

---

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/17095 

Problem Summary: 
When tidb-server is killed by system oom-killer, we have no information to investigate which SQL is suspect that consumes unexpected memory and causes process down.

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed: 
1. Add a config Performance.ServerMemoryAlert to set the threshold manually.
2. If Performance.ServerMemoryQuota is set, use instance memory usage and ServerMemoryQuota * 80% to check oom risk.
3. If Performance.ServerMemoryQuota is not set, use system memory usage and total memory * 80% to check oom risk.

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- Record SQLs and heap profile when memUsage is more than 80% of memQuota

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
@wshwsh12 conflicts need to be resolved

**Review by XuHuaiyu — COMMENTED** [review]
LGTM

**Review by lzmhhh123 — DISMISSED** [review]
LGTM

**Comment by wshwsh12** [comment]
Maybe need cherry-pick this pr .... https://github.com/pingcap/tidb/pull/17532

**Review by XuHuaiyu — APPROVED** [review]
LGTM

File: util/expensivequery/memory_usage_alarm.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+	if alert := config.GetGlobalConfig().Performance.MemoryUsageAlarmRatio; alert == 0 || alert == 1 {
+		record.err = errors.New("close memory usage alarm recorder")
+		return
+	}
+	if quota := config.GetGlobalConfig().Performance.ServerMemoryQuota; quota != 0 {
+		record.serverMemoryQuota = quota
+		record.isServerMemoryQuotaSet = true
+	} else {
+		// TODO: Get the memory info in container directly.
+		record.serverMemoryQuota, record.err = memory.MemTotal()
```
Comment (breeswish) [inline util/expensivequery/memory_usage_alarm.go:59]: How abouot container environment?

Comment (wshwsh12) [reply util/expensivequery/memory_usage_alarm.go:59]: Now the only way to use it is to set server-memory-quota config.

**Pull request merged**